### PR TITLE
Cast tensor dtype according to the schema in a MaxCompute table

### PR DESCRIPTION
**Pull request by workingloong**
In dataset_fn, we can cast the tensor to dtype according to the column type in a MaxCompute table.

### CODE REVIEW

File: elasticdl/python/data/reader/data_reader.py [hunk 1]
```diff
@@ -1,11 +1,33 @@
...
+        """Get TensorFlow dtype according to the column name in
+
+        Args:
+            column_name: The column name in a MaxCompute table.
+
+        Returns:
+            TensorFlow dtype
+        """
+        maxcompute_dtype = self.column_dtypes[
+            self.column_names.index(column_name)
```
Comment (chunyang-wen) [inline elasticdl/python/data/reader/data_reader.py]: Add a check for `column_dtypes` ? It can be None.

Comment (chunyang-wen) [reply elasticdl/python/data/reader/data_reader.py]: `index` is `O(n)`.  You can use a dict, keys are names, values are types? And add `column_names` and `column_types` as a property.

Comment (workingloong) [reply elasticdl/python/data/reader/data_reader.py]: Good! I have fixed `column_dtypes` to a dict.

File: elasticdl/python/data/reader/data_reader.py [hunk 2]
```diff
@@ -1,11 +1,47 @@
...
+
+    Attributes:
+        column_names: A list with column names
+        column_dtypes: A dict where the key is a column name
+            and the value is a dtype
+    """
+
     def __init__(self, column_names, column_dtypes=None):
         self.column_names = column_names
         self.column_dtypes = column_dtypes
```
Comment (chunyang-wen) [inline elasticdl/python/data/reader/data_reader.py:18]: When do we use `column_names`? And following code is FYI.

```python
@property
def column_dtypes(self):
    return self.column_dtypes

@column_dtypes.setter
def column_dtypes(self, col_dtypes):
    self.column_dtypes = col_dtypes
```

Comment (workingloong) [reply elasticdl/python/data/reader/data_reader.py:18]: We use `column_names` to create a dataset  in `dataset_fn`. The features of the datset is a dict like {column_name: tensor}.

File: elasticdl/python/data/reader/data_reader.py [hunk 3]
```diff
@@ -1,11 +1,51 @@
...
 
+from elasticdl.python.common.dtypes import MAXCOMPUTE_DTYPE_TO_TF
+
 
 class Metadata(object):
+    """ Metadata of a dataset containing column name and dtype
+
+    Attributes:
+        column_names: A list with column names
+        column_dtypes: A dict where the key is a column name
```
Comment (brightcoder01) [inline elasticdl/python/data/reader/data_reader.py:11]: the value is dtype of MaxCompute or tf?

File: elasticdl/python/data/reader/odps_reader.py [hunk 4]
```diff
@@ -29,10 +29,10 @@ def _init_metadata(self):
             )
 
         if self._metadata.column_names:
-            self._metadata.column_dtypes = [
-                table_schema[column_name].type
-                for column_name in self._metadata.column_names
-            ]
+            column_dtypes = {}
+            for column_name in self._metadata.column_names:
```
Comment (brightcoder01) [inline elasticdl/python/data/reader/odps_reader.py]: column_dtypes = {column_name: table_schema[column_name].type for column_name in self._metadata.column_names}

File: elasticdl/python/common/dtypes.py [hunk 5]
```diff
@@ -38,3 +39,10 @@ def is_numpy_dtype_allowed(dtype):
     np.float64: types_pb2.DT_DOUBLE,
     np.bool: types_pb2.DT_BOOL,
 }
+
+# TODO: There are many dtypes in MaxCompute and we can add them if needed.
+MAXCOMPUTE_DTYPE_TO_TF = {
```
Comment (brightcoder01) [inline elasticdl/python/common/dtypes.py]: MAXCOMPUTE_DTYPE_TO_TF_DTYPE

File: elasticdl/python/data/reader/data_reader.py [hunk 6]
```diff
@@ -1,11 +1,51 @@
...
 
+    @property
+    def column_dtypes(self):
+        return self.__column_dtypes
+
+    @column_dtypes.setter
+    def column_dtypes(self, column_dtypes):
+        self.__column_dtypes = column_dtypes
+
+    def get_tf_dtype_by_maxcompute(self, column_name):
```
Comment (brightcoder01) [inline elasticdl/python/data/reader/data_reader.py]: get_tf_dtype_from_maxcompute_column

File: elasticdl/python/data/reader/data_reader.py (hunk 3 above)
Comment (workingloong) [reply elasticdl/python/data/reader/data_reader.py:11]: A dtype of MaxCompute or pandas for CSV file, but not a TF dtype. I will add it into the docstring.

**Review by brightcoder01 — APPROVED** [review]

**Pull request merged**